src/System.Reflection/tests/AssemblyTests.cs

```diff
@@ -0,0 +1,342 @@
+        }
+
+        [Theory]
+        [MemberData(nameof(CreateInstance_TestData))]
+        public void CreateInstance(Assembly assembly, string typeName, Type expectedType)
+        {
+            Assert.IsType(expectedType, assembly.CreateInstance(typeName));
+            Assert.IsType(expectedType, assembly.CreateInstance(typeName, false));
+            Assert.IsType(expectedType, assembly.CreateInstance(typeName, true));
+
+            Assert.IsType(expectedType, assembly.CreateInstance(typeName.ToUpper(), true));
+            Assert.IsType(expectedType, assembly.CreateInstance(typeName.ToLower(), true));
+        }
+
+        public static IEnumerable<object[]> CreateInstance_Invalid_TestData()
+        {
+            yield return new object[] { "", typeof(ArgumentException) };
+            yield return new object[] { null, typeof(ArgumentNullException) };
+            yield return new object[] { typeof(AssemblyClassWithPrivateCtor).FullName, typeof(MissingMethodException) };
+            yield return new object[] { typeof(AssemblyClassWithNoDefaultCtor).FullName, typeof(MissingMethodException) };
+        }
+
+        [Theory]
+        [MemberData(nameof(CreateInstance_Invalid_TestData))]
+        public void CreateInstance_Invalid(string typeName, Type exceptionType)
+        {
+            Assembly assembly = Helpers.ExecutingAssembly;
+            Assert.Throws(exceptionType, () => Helpers.ExecutingAssembly.CreateInstance(typeName));
+            Assert.Throws(exceptionType, () => Helpers.ExecutingAssembly.CreateInstance(typeName, true));
+            Assert.Throws(exceptionType, () => Helpers.ExecutingAssembly.CreateInstance(typeName, false));
+        }
+
+        [Fact]
+        public void CreateQualifiedName()
+        {
+            string assemblyName = Helpers.ExecutingAssembly.ToString();
+            Assert.Equal(typeof(AssemblyTests).FullName + ", " + assemblyName, Assembly.CreateQualifiedName(assemblyName, typeof(AssemblyTests).FullName));
+        }
+
+        [Fact]
+        public void GetReferencedAssemblies()
+        {
+            // It is too brittle to depend on the assembly references so we just call the method and check that it does not throw.
+            AssemblyName[] assemblies = Helpers.ExecutingAssembly.GetReferencedAssemblies();
+            Assert.NotEmpty(assemblies);
+        }
+
+        public static IEnumerable<object[]> Modules_TestData()
+        {
+            yield return new object[] { LoadSystemCollectionsAssembly() };
+            yield return new object[] { LoadSystemReflectionAssembly() };
+        }
+
+        [Theory]
+        [MemberData(nameof(Modules_TestData))]
+        public void Modules(Assembly assembly)
+        {
+            foreach (Module module in assembly.Modules)
+            {
+                Assert.NotNull(module);
+            }
```

Comment:
The previous version of this test also verifies that assembly.Modules isn't empty (if it were empty, the call to Current would fail or return null).